soc/arm/nxp_imx/rt5xx/power.c

```diff
@@ -0,0 +1,163 @@
+#include "fsl_power.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
+
+/*!< Power down all unnecessary blocks */
+#define NODE_ID DT_INST(0, nxp_pdcfg_power)
+#define APP_EXCLUDE_FROM_DEEPSLEEP ((const uint32_t[]) \
+					DT_PROP_OR(NODE_ID, deep_sleep_config, {}))
+
+#define APP_EXCLUDE_FROM_DEEP_POWERDOWN ((const uint32_t[]){0, 0, 0, 0})
+
+/* System clock frequency. */
+extern uint32_t SystemCoreClock;
+
+#if CONFIG_XIP
+/*
+ * Declare pin configuration state for flexspi pins.
+ * These pins will be reconfigured for power savings
+ */
+#define PINCTRL_STATE_XIP_SLEEP PINCTRL_STATE_PRIV_START
+#define FLEXSPI_DEVICE DT_NODELABEL(flexspi)
+PINCTRL_DT_DEFINE(FLEXSPI_DEVICE);
+const struct pinctrl_dev_config *flexspi_pincfg = PINCTRL_DT_DEV_CONFIG_GET(FLEXSPI_DEVICE);
+const struct pinctrl_state *flexspi_pins_default_state;
+const struct pinctrl_state *flexspi_pins_sleep_state;
+/* Flag indicates if we can reconfigure flexspi pins for additional power savings */
+static bool flexspi_pin_reconfig;
+#endif
+
+static uint32_t isp_pin[3];
+
+__ramfunc void set_deepsleep_pin_config(void)
+{
+	/* Backup Pin configuration. */
+	isp_pin[0] = IOPCTL->PIO[1][15];
+	isp_pin[1] = IOPCTL->PIO[3][28];
+	isp_pin[2] = IOPCTL->PIO[3][29];
+
+	/* Disable ISP Pin pull-ups and input buffers to avoid current leakage */
+	IOPCTL->PIO[1][15] = 0;
+	IOPCTL->PIO[3][28] = 0;
+	IOPCTL->PIO[3][29] = 0;
+#if CONFIG_XIP
+	/* Apply flexspi pinctrl state for Sleep state if defined through device tree */
+	if (flexspi_pin_reconfig) {
+		pinctrl_configure_pins(flexspi_pins_sleep_state->pins,
+				       flexspi_pins_sleep_state->pin_cnt,
+				       PINCTRL_REG_NONE);
+	}
+#endif
+}
+
+__ramfunc void restore_deepsleep_pin_config(void)
+{
+	/* Restore the Pin configuration. */
+	IOPCTL->PIO[1][15] = isp_pin[0];
+	IOPCTL->PIO[3][28] = isp_pin[1];
+	IOPCTL->PIO[3][29] = isp_pin[2];
+#if CONFIG_XIP
+	/* Apply flexspi pinctrl state for default state */
+	if (flexspi_pin_reconfig) {
+		pinctrl_configure_pins(flexspi_pins_default_state->pins,
+				       flexspi_pins_default_state->pin_cnt,
+				       PINCTRL_REG_NONE);
+	}
+#endif
+}
+
+/* Invoke Low Power/System Off specific Tasks */
+__weak void pm_state_set(enum pm_state state, uint8_t substate_id)
+{
+	ARG_UNUSED(substate_id);
+
+	/* FIXME: When this function is entered the Kernel has disabled
+	 * interrupts using BASEPRI register. This is incorrect as it prevents
+	 * waking up from any interrupt which priority is not 0. Work around the
+	 * issue and disable interrupts using PRIMASK register as recommended
+	 * by ARM.
+	 */
+
+	/* Set PRIMASK */
+	__disable_irq();
+
+	/* Set BASEPRI to 0 */
+	irq_unlock(0);
+
+	switch (state) {
+	case PM_STATE_RUNTIME_IDLE:
+		POWER_EnterSleep();
+		break;
+	case PM_STATE_SUSPEND_TO_IDLE:
+		set_deepsleep_pin_config();
+		POWER_EnterDeepSleep(APP_EXCLUDE_FROM_DEEPSLEEP);
+		restore_deepsleep_pin_config();
+		break;
+	case PM_STATE_SOFT_OFF:
```

Comment:
Is this state defined in DT ? I just saw runtime-idle and suspend-to-idle.

Reply:
It is not. For putting this in DT I will have to add dummy residency values as we not enter these states from the idle loop.
This state should be entered only when the user application decides to put the platform in SOFT_OFF, recovery from this state causes a `warm-reset`.
This is also similar to other implementations of `PM_STATE_SOFT_OFF`